src/include/neorados/RADOS.hpp

```diff
@@ -213,53 +213,53 @@ class Op {
   Op& operator =(Op&&);
   ~Op();
 
-  void set_excl();
-  void set_failok();
-  void set_fadvise_random();
-  void set_fadvise_sequential();
-  void set_fadvise_willneed();
-  void set_fadvise_dontneed();
-  void set_fadvise_nocache();
-
-  void cmpext(uint64_t off, ceph::buffer::list&& cmp_bl, std::size_t* s);
-  void cmpxattr(std::string_view name, cmpxattr_op op,
+  Op&& set_excl();
+  Op&& set_failok();
+  Op&& set_fadvise_random();
+  Op&& set_fadvise_sequential();
+  Op&& set_fadvise_willneed();
+  Op&& set_fadvise_dontneed();
+  Op&& set_fadvise_nocache();
+
+  Op&& cmpext(uint64_t off, ceph::buffer::list&& cmp_bl, std::size_t* s);
+  Op&& cmpxattr(std::string_view name, cmpxattr_op op,
 		const ceph::buffer::list& val);
-  void cmpxattr(std::string_view name, cmpxattr_op op, std::uint64_t val);
-  void assert_version(uint64_t ver);
-  void assert_exists();
-  void cmp_omap(const boost::container::flat_map<
-		  std::string,
-		  std::pair<ceph::buffer::list, int>>& assertions);
-
-  void exec(std::string_view cls, std::string_view method,
+  Op&& cmpxattr(std::string_view name, cmpxattr_op op, std::uint64_t val);
+  Op&& assert_version(uint64_t ver);
+  Op&& assert_exists();
+  Op&& cmp_omap(const boost::container::flat_map<std::string,
+		                                 std::pair<ceph::buffer::list,
+		                                           int>>& assertions);
+
+  Op&& exec(std::string_view cls, std::string_view method,
```

Comment:
So we can feed the output right into `execute`, which takes its argument by rvalue

(execute is destructive to operations in this library and in vanilla rados as well, I made execute move from the op to reflect that.)

Reply:
i like that `execute()` takes the op by rvalue-ref, and think it works well without chaining:
```c++
neorados::WriteOp op;
op.create(true);
op.write_full(std::move(bl));
co_await rados.execute(obj, ioc, std::move(op), use_awaitable);
```
so i'm trying to picture the style you want to encourage here, especially as it scales to composite ops
```c++
// 1: chain all calls from an rvalue op
co_await rados.execute(obj, ioc, neorados::WriteOp{}.create(true).write_full(std::move(bl)), use_awaitable);

// 2: chain all calls from an lvalue op
neorados::WriteOp op;
co_await rados.execute(obj, ioc, op.create(true).write_full(std::move(bl)), use_awaitable);

// 3: chain the last call from an lvalue op
neorados::WriteOp op;
op.create(true);
co_await rados.execute(obj, ioc, op.write_full(std::move(bl)), use_awaitable);
```
1 is ugly, but it does avoid leaving a moved-from variable on the stack. i don't like how 2 and 3 obscure the fact that `execute()` is destructive, since they both move from `op` without the caller ever typing `std::move(op)`. you might clarify that by only enabling the rvalue-chaining only for rvalue overloads (ie `Op&& exec(...) &&;`), but i think we'd still want overloads for lvalue ref too